### ISSUE TITLE
17.05 : Lot of  Jobs fail due to missing metadata

### ISSUE THREAD

**Issue by juliechevalier** [galaxyproject/galaxy#4724]
Hi,
I have update my galaxy instance to 17.05, on the last commit (re-updated this week 2017/09/25).
So since this update, I have lot of bug regarding metadata. Some job fail while they run correctly but the status is error because of problem with metadata at the end of the job. 
the error message can be different according tool, for example for trim tool from fastx-toolkit, it's : [Erno2] No such file or directory ".../database/tmp/tmp..."
With MACS2 from toolshed it's just : Fatal Error : exit code 1 ()
I have had this problem with different tools installed from toolshed or native, and it happens randomly.
I've tested the protocol with "click on pencil and click on auto-detect", but the jobs stay in yellow and their is no change.
It's untractable with lot of data and if we run tool on multiple datasets!

Any idea or any help to return to the previous version of galaxy without crash my galaxy?
Thanks
Julie

**Comment by mvdbeek** [comment]
Please include the error message that you are seeing.

**Comment by juliechevalier** [comment]
For MACS2 the bug give me this message :
Tool execution generated the following error message:
```
Fatal error: Exit code 1 ()

The tool produced the following additional output:

INFO  @ Thu, 28 Sep 2017 19:35:03: 
# Command line: callpeak --name MACS2 -t /data/galaxy-dist/database/files/080/dataset_80551.dat -c /data/galaxy-dist/database/files/080/dataset_80457.dat --format=BAM --gsize 2150570000 --bw=300 --ratio 1.0 --slocal 1000 --llocal 10000 --keep-dup 1 --bdg --pvalue 0.001
# ARGUMENTS LIST:
# name = MACS2
# format = BAM
# ChIP-seq file = ['/data/galaxy-dist/database/files/080/dataset_80551.dat']
# control file = ['/data/galaxy-dist/database/files/080/dataset_80457.dat']
# effective genome size = 2.15e+09
# band width = 300
# model fold = [5, 50]
# pvalue cutoff = 1.00e-03
# qvalue will not be calculated and reported as -1 in the final output.
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 1000 bps and 10000 bps
# Broad region calling is off
....
#Done!
```
And if I click on the eye, I can see data !!! But the assignment of columns are strange and not good because of metadata problem I think.

**Comment by juliechevalier** [comment]
For the Trim tool I have this message :
Traceback...
IOError: [Errno 2] No such file or directory: '/galaxy/database/tmp/tmpleXN4b

But again I can see the result of the tool run.  Strangely this error message appeared in very old version of Galaxy when galaxy was restarted while job were in running.
But it's not my case here. I haven't restart Galaxy

**Comment by juliechevalier** [comment]
I'm trying to isolate the problem. And I have 2 more information :
1. When I restart the job which failed for metadata issue, it runs correctly and its status becomes OK !
2. I believe that such random error appears only when we run a tool with the option "multiple datasets". So in this case I can run several time the tool on multiple datasets, at each run there are some job which failed but not the same between run (it's not a problem link to a particular dataset). But I have never seen this bug when I run the tool on each dataset separately.

Thanks for your help

**Comment by mvdbeek** [comment]
Maybe we can start with the metadata problem. When this outright fails (it should be a red dataset in the history), what do you see in the log files of galaxy ?

**Comment by juliechevalier** [comment]
The log file is very huge, so I've been tried to rerun exaclty the same job in the same conditions : 15 datasets at the same time for Trim tool...
And all jobs in history appear OK !!!
I don't understand.
I can attach you the corresponding which comprises some warning or error but there is no problem in the history this time

**Comment by juliechevalier** [comment]
file:///home/julie/Bureau/galaxy_log_trim.txt

**Comment by mvdbeek** [comment]
One trick to get to the right portion of the log (assuming you're the admin) is to search through the log file using the output path of the job that failed ("Full Path:   ") that should be shown if you click on the "i" icon in the history that contains the failed job.

**Comment by juliechevalier** [comment]
Ok, Thanks!
I've found this for a job launch on bowtie with the same error :

galaxy.jobs.command_factory INFO 2017-09-25 15:59:32,840 Built script [/data/galaxy-dist/database/job_working_directory/037/37906/tool_script.sh] for tool command [export PATH=$PATH:'/data/galaxy-dist/tool_dependency/_conda/envs/__bowtie@1.2.0/bin' ; python /data/galaxy-dist/tools/sr_mapping/bowtie_wrapper_v1.2.0.py --threads="2" --output=/data/galaxy-dist/database/files/079/dataset_79961.dat --galaxy_input_format="fastqsanger" --formatInput=fastq --dataType="solexa" --suppressHeader=false --genomeSource=indexed --ref="/home/galaxy/Data_Storage/Galaxy_tool-data/indexes/mm10/bowtie/base/mm10" --paired=single --input1=/data/galaxy-dist/database/files/079/dataset_79922.dat --params=full --skip=0 --alignLimit=-1 --trimH=0 --trimL=0 --mismatchSeed=2 --mismatchQual=70 --seedLen=28 --rounding=round --maqSoapAlign=-1 --tryHard=doTryHard --valAlign=1 --allValAligns=noAllValAligns --suppressAlign=1 --best=doBest --maxBacktracks=800 --strata=doStrata --offrate=-1 --seed=-1 ;/bin/bash /data/galaxy-dist/database/job_working_directory/037/37906/tmpe7J3jS]
galaxy.tools.deps DEBUG 2017-09-25 15:59:32,904 Using dependency samtools version None of type conda
galaxy.jobs.runners DEBUG 2017-09-25 15:59:32,906 (37906) command is: rm -rf working; mkdir -p working; cd working; /data/galaxy-dist/database/job_working_directory/037/37906/tool_script.sh; return_code=$?; cd '/data/galaxy-dist/database/job_working_directory/037/37906'; 
if [ "$GALAXY_LIB" != "None" ]; then
    if [ -n "$PYTHONPATH" ]; then
        PYTHONPATH="$GALAXY_LIB:$PYTHONPATH"
    else
        PYTHONPATH="$GALAXY_LIB"
    fi
    export PYTHONPATH
fi
if [ "$GALAXY_VIRTUAL_ENV" != "None" -a -z "$VIRTUAL_ENV"      -a -f "$GALAXY_VIRTUAL_ENV/bin/activate" ]; then
    . "$GALAXY_VIRTUAL_ENV/bin/activate"
fi
GALAXY_PYTHON=`command -v python`
export PATH=$PATH:'/data/galaxy-dist/tool_dependency/_conda/envs/__samtools@1.3.1/bin' ; python "/data/galaxy-dist/database/job_working_directory/037/37906/set_metadata_4ARlcK.py" "/data/galaxy-dist/database/tmp/tmpnSECrk" "/data/galaxy-dist/database/job_working_directory/037/37906/working/galaxy.json" "/data/galaxy-dist/database/job_working_directory/037/37906/metadata_in_HistoryDatasetAssociation_71500_hNlBKw,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_kwds_HistoryDatasetAssociation_71500_BjZMYe,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_out_HistoryDatasetAssociation_71500_XgH8Ga,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_results_HistoryDatasetAssociation_71500_KGM0dG,/data/galaxy-dist/database/files/079/dataset_79961.dat,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_override_HistoryDatasetAssociation_71500_gCb7or" 5242880; sh -c "exit $return_code"
galaxy.jobs.runners.local DEBUG 2017-09-25 15:59:32,938 (37906) executing job script: /data/galaxy-dist/database/job_working_directory/037/37906/galaxy_37906.sh
galaxy.jobs DEBUG 2017-09-25 15:59:32,969 (37906) Persisting job destination (destination id: normal_longjobs)
192.168.100.175 - - [25/sept./2017:15:59:33 +0200] "GET /galaxy/api/histories/13c61b570ff1acdd/contents?details=700c36128f6b5177%2Ca3f0f866bc3997ba%2C06715b5167be6a79%2Cd2d40db7396af371%2C6267059a8f222e7e%2C1f558c9ce190b484%2Cca745ade9815fa10&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2017-09-25T13%3A59%3A29.000Z&qv=False&qv=False HTTP/1.1" 200 - "http://192.168.100.98:8080/galaxy?tool_id=bowtie_wrapper120&version=1-1.2.0&__identifer=lt307q1kko" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:55.0) Gecko/20100101 Firefox/55.0"
...
galaxy.jobs.runners.local DEBUG 2017-09-25 16:16:28,298 execution finished: /data/galaxy-dist/database/job_working_directory/037/37906/galaxy_37906.sh
192.168.100.175 - - [25/sept./2017:16:16:28 +0200] "GET /galaxy/api/histories/13c61b570ff1acdd/contents?details=700c36128f6b5177%2Ca3f0f866bc3997ba%2C06715b5167be6a79%2Cd2d40db7396af371%2C6267059a8f222e7e%2C1f558c9ce190b484%2Cca745ade9815fa10%2C9b76aaa36a56c0e5&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2017-09-25T14%3A16%3A24.000Z&qv=False&qv=False HTTP/1.1" 200 - "http://192.168.100.98:8080/galaxy?tool_id=bowtie_wrapper120&version=1-1.2.0&__identifer=lt307q1kko" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:55.0) Gecko/20100101 Firefox/55.0"
galaxy.jobs.output_checker DEBUG 2017-09-25 16:16:28,303 Tool produced standard error failing job - [Traceback (most recent call last):
  File "/data/galaxy-dist/database/job_working_directory/037/37906/set_metadata_4ARlcK.py", line 1, in <module>
    from galaxy_ext.metadata.set_metadata import set_metadata; set_metadata()
  File "/data/galaxy-dist]
galaxy.jobs DEBUG 2017-09-25 16:16:28,775 (37906) setting dataset 79961 state to ERROR
galaxy.util WARNING 2017-09-25 16:16:28,848 Unable to honor umask (02) for /data/galaxy-dist/database/files/079/dataset_79961.dat, tried to set: 0664 but mode remains 0777, error was: [Errno 1] Operation not permitted: '/data/galaxy-dist/database/files/079/dataset_79961.dat'
galaxy.util WARNING 2017-09-25 16:16:28,848 Unable to honor primary group (grp.struct_group(gr_name='galaxy', gr_passwd='x', gr_gid=1001, gr_mem=[])) for /data/galaxy-dist/database/files/079/dataset_79961.dat, group remains grp.struct_group(gr_name='users', gr_passwd='x', gr_gid=100, gr_mem=[]), error was: [Errno 1] Operation not permitted: '/data/galaxy-dist/database/files/079/dataset_79961.dat'

**Comment by juliechevalier** [comment]
Ok thanks,
I've found this for a job launched on bowtie failing with the same error :
```
galaxy.jobs.command_factory INFO 2017-09-25 15:59:32,840 Built script [/data/galaxy-dist/database/job_working_directory/037/37906/tool_script.sh] for tool command [export PATH=$PATH:'/data/galaxy-dist/tool_dependency/_conda/envs/__bowtie@1.2.0/bin' ; python /data/galaxy-dist/tools/sr_mapping/bowtie_wrapper_v1.2.0.py --threads="2" --output=/data/galaxy-dist/database/files/079/dataset_79961.dat --galaxy_input_format="fastqsanger" --formatInput=fastq --dataType="solexa" --suppressHeader=false --genomeSource=indexed --ref="/home/galaxy/Data_Storage/Galaxy_tool-data/indexes/mm10/bowtie/base/mm10" --paired=single --input1=/data/galaxy-dist/database/files/079/dataset_79922.dat --params=full --skip=0 --alignLimit=-1 --trimH=0 --trimL=0 --mismatchSeed=2 --mismatchQual=70 --seedLen=28 --rounding=round --maqSoapAlign=-1 --tryHard=doTryHard --valAlign=1 --allValAligns=noAllValAligns --suppressAlign=1 --best=doBest --maxBacktracks=800 --strata=doStrata --offrate=-1 --seed=-1 ;/bin/bash /data/galaxy-dist/database/job_working_directory/037/37906/tmpe7J3jS]
galaxy.tools.deps DEBUG 2017-09-25 15:59:32,904 Using dependency samtools version None of type conda
galaxy.jobs.runners DEBUG 2017-09-25 15:59:32,906 (37906) command is: rm -rf working; mkdir -p working; cd working; /data/galaxy-dist/database/job_working_directory/037/37906/tool_script.sh; return_code=$?; cd '/data/galaxy-dist/database/job_working_directory/037/37906'; 
if [ "$GALAXY_LIB" != "None" ]; then
    if [ -n "$PYTHONPATH" ]; then
        PYTHONPATH="$GALAXY_LIB:$PYTHONPATH"
    else
        PYTHONPATH="$GALAXY_LIB"
    fi
    export PYTHONPATH
fi
if [ "$GALAXY_VIRTUAL_ENV" != "None" -a -z "$VIRTUAL_ENV"      -a -f "$GALAXY_VIRTUAL_ENV/bin/activate" ]; then
    . "$GALAXY_VIRTUAL_ENV/bin/activate"
fi
GALAXY_PYTHON=`command -v python`
export PATH=$PATH:'/data/galaxy-dist/tool_dependency/_conda/envs/__samtools@1.3.1/bin' ; python "/data/galaxy-dist/database/job_working_directory/037/37906/set_metadata_4ARlcK.py" "/data/galaxy-dist/database/tmp/tmpnSECrk" "/data/galaxy-dist/database/job_working_directory/037/37906/working/galaxy.json" "/data/galaxy-dist/database/job_working_directory/037/37906/metadata_in_HistoryDatasetAssociation_71500_hNlBKw,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_kwds_HistoryDatasetAssociation_71500_BjZMYe,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_out_HistoryDatasetAssociation_71500_XgH8Ga,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_results_HistoryDatasetAssociation_71500_KGM0dG,/data/galaxy-dist/database/files/079/dataset_79961.dat,/data/galaxy-dist/database/job_working_directory/037/37906/metadata_override_HistoryDatasetAssociation_71500_gCb7or" 5242880; sh -c "exit $return_code"
galaxy.jobs.runners.local DEBUG 2017-09-25 15:59:32,938 (37906) executing job script: /data/galaxy-dist/database/job_working_directory/037/37906/galaxy_37906.sh
galaxy.jobs DEBUG 2017-09-25 15:59:32,969 (37906) Persisting job destination (destination id: normal_longjobs)
192.168.100.175 - - [25/sept./2017:15:59:33 +0200] "GET /galaxy/api/histories/13c61b570ff1acdd/contents?details=700c36128f6b5177%2Ca3f0f866bc3997ba%2C06715b5167be6a79%2Cd2d40db7396af371%2C6267059a8f222e7e%2C1f558c9ce190b484%2Cca745ade9815fa10&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2017-09-25T13%3A59%3A29.000Z&qv=False&qv=False HTTP/1.1" 200 - "http://192.168.100.98:8080/galaxy?tool_id=bowtie_wrapper120&version=1-1.2.0&__identifer=lt307q1kko" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:55.0) Gecko/20100101 Firefox/55.0"
...
galaxy.jobs.runners.local DEBUG 2017-09-25 16:16:28,298 execution finished: /data/galaxy-dist/database/job_working_directory/037/37906/galaxy_37906.sh
192.168.100.175 - - [25/sept./2017:16:16:28 +0200] "GET /galaxy/api/histories/13c61b570ff1acdd/contents?details=700c36128f6b5177%2Ca3f0f866bc3997ba%2C06715b5167be6a79%2Cd2d40db7396af371%2C6267059a8f222e7e%2C1f558c9ce190b484%2Cca745ade9815fa10%2C9b76aaa36a56c0e5&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2017-09-25T14%3A16%3A24.000Z&qv=False&qv=False HTTP/1.1" 200 - "http://192.168.100.98:8080/galaxy?tool_id=bowtie_wrapper120&version=1-1.2.0&__identifer=lt307q1kko" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:55.0) Gecko/20100101 Firefox/55.0"
galaxy.jobs.output_checker DEBUG 2017-09-25 16:16:28,303 Tool produced standard error failing job - [Traceback (most recent call last):
  File "/data/galaxy-dist/database/job_working_directory/037/37906/set_metadata_4ARlcK.py", line 1, in <module>
    from galaxy_ext.metadata.set_metadata import set_metadata; set_metadata()
  File "/data/galaxy-dist]
galaxy.jobs DEBUG 2017-09-25 16:16:28,775 (37906) setting dataset 79961 state to ERROR
galaxy.util WARNING 2017-09-25 16:16:28,848 Unable to honor umask (02) for /data/galaxy-dist/database/files/079/dataset_79961.dat, tried to set: 0664 but mode remains 0777, error was: [Errno 1] Operation not permitted: '/data/galaxy-dist/database/files/079/dataset_79961.dat'
galaxy.util WARNING 2017-09-25 16:16:28,848 Unable to honor primary group (grp.struct_group(gr_name='galaxy', gr_passwd='x', gr_gid=1001, gr_mem=[])) for /data/galaxy-dist/database/files/079/dataset_79961.dat, group remains grp.struct_group(gr_name='users', gr_passwd='x', gr_gid=100, gr_mem=[]), error was: [Errno 1] Operation not permitted: '/data/galaxy-dist/database/files/079/dataset_79961.dat'
```

**Comment by mvdbeek** [comment]
Hmm, OK, and does this jobs' stderr contain a message ?

**Comment by juliechevalier** [comment]
Yes the same than with the bug icon and I've reported for metadata:
```
Traceback (most recent call last):
  File "/data/galaxy-dist/database/job_working_directory/037/37906/set_metadata_4ARlcK.py", line 1, in <module>
    from galaxy_ext.metadata.set_metadata import set_metadata; set_metadata()
  File "/data/galaxy-dist/lib/galaxy_ext/metadata/set_metadata.py", line 85, in set_metadata
    datatypes_registry.load_datatypes( root_dir=galaxy_root, config=datatypes_config )
  File "/data/galaxy-dist/lib/galaxy/datatypes/registry.py", line 104, in load_datatypes
    tree = galaxy.util.parse_xml( config )
  File "/data/galaxy-dist/lib/galaxy/util/__init__.py", line 214, in parse_xml
    root = tree.parse( fname, parser=ElementTree.XMLParser( target=DoctypeSafeCallbackTarget() ) )
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 647, in parse
    source = open(source, "rb")
IOError: [Errno 2] No such file or directory: '/data/galaxy-dist/database/tmp/tmpnSECrk'

```

**Comment by mvdbeek** [comment]
> Yes the same than with the bug icon and I've reported for metadata:

In your initial report you've cut the output, so I couldn't tell exactly what is going on.

> File "/data/galaxy-dist/lib/galaxy/datatypes/registry.py", line 104, in load_datatypes

So what happened here (most likely) is that a job wrote the datatype registry to /data/galaxy-dist/database/tmp/tmpnSECrk, which got removed when galaxy was restarted. In one of the recent commits of 17.05 we changed how the datatype registry which you need for setting metadata is written out (this now ends up the job working directory for greater isolation between jobs and and galaxy state). So when your job finished, this file didn't exist, and the metadata setting failed. We shouldn't have included such a change in an existing release without more testing, that is entirely our fault.
The bright side is that this particular error should not occur again.

**Comment by juliechevalier** [comment]
OK, thanks
I hope this bug will be corrected soon. For the moment with the latest release, it still occurs.
And another strnage thing is that in my case, we have NOT restarted Galaxy between the start of the job and the end of the job.

Thanks again for your help and your time.

Julie

**Comment by mvdbeek** [comment]
> For the moment with the latest release, it still occurs.

I thought this isn't occuring anymore ?

**Comment by abretaud** [comment]
Hey,
Adding some noise here: we have some similar problem on our galaxy instance (17.05, 10efd7fac6e38c5ca07d94c351d10473811c7d6c)
I'll see if updating to 17.09 fixes this problem, but just in case, some info:
- the error seems to happen randomly on some jobs which become red with the same kind of error (even if the content is correct)
- we don't restart the instance during the job's life

I guess for some reason the temp file is not created, or is deleted.

I'll update this if I still get it after 17.09 update (don't know when...)

**Comment by mvdbeek** [comment]
@abretaud there were quite some fixes on 17.05 since 10efd7f, notably 54357f38e9929f3f3a084f298bd644395976a91a which may fix what you are describing

**Comment by natefoo** [comment]
I believe this issue has been fixed by @mvdbeek in the linked commits, but if you are still having trouble, please feel free to reply and we can reopen the issue.
